### ISSUE TITLE
Types not checked. Apollo typed query hook with Hasura endpoint allows anything. 

### ISSUE THREAD

**Issue by seanonthenet** [graphql-editor/graphql-zeus#232]
Is this expected behaviour?

Autocomplete works but the query essentially allows any object without error:

<img width="445" alt="Screenshot 2021-11-11 at 13 12 03@2x" src="https://user-images.githubusercontent.com/4939629/141355378-c8442ad0-7096-4ec5-b20f-0b0c2582292c.png">

Love Zeus. ❤️⚡️ Thanks.

**Comment by aexol** [comment]
I think we can repair it thanks!!

**Comment by IceBlizz6** [comment]
Discovered this issue yesterday while testing zeus.
It seems like it is only possible to add non-existing fields if there are at least 1 existing field declared in the query.

So as an example:

=== Schema ===
```
type MalSeasonIdentifier  {
  season: MalYearSeason! 
  year: Int! 
}

type Query {
  animeSeasonList: [MalSeasonIdentifier!]! 
}
```

=== TS ===
this does not give any errors as 'year' is a valid field ->
```
const result = await chain("query")({
	animeSeasonList: {
		year: true,
		fieldThatDoesNotExist: "f"
	}
})
```

this does give error ->
```
const result = await chain("query")({
	animeSeasonList: {
		fieldThatDoesNotExist: "f"
	}
})
```
![image](https://user-images.githubusercontent.com/38980218/146659751-c4326fad-ff76-4a6f-9fe1-7eeac67d70a2.png)

Having a safety mechanism to alert us of incorrect queries is important, especially as system evolves and fields may get deprecated and disappear.

Keep up the good work, i hope someone will be able to look into this soon.

**Comment by omaiboroda** [comment]
Any ideas how this may be fixed? Pretty fundamental for error-avoiding

**Comment by aexol** [comment]
play with types and try to find a solution. Maybe with latest version of TS you can do smth, before it was impossible for me to do

**Comment by karibertils** [comment]
Also wondering if it's possible to fix this ?

**Comment by aexol** [comment]
@karibertils https://zeus.graphqleditor.com/page/plugins/typedDocumentNode.html we are in the process of transitioning to this. Try using this way!

**Comment by karibertils** [comment]
@aexol 

This looks like a nice way to do things, and I see the types are now working for the variables which is really cool. But typescript still shows no error when adding random fields which don't exist in the schema. See below.

![image](https://user-images.githubusercontent.com/5122320/178154733-479d221f-72f7-4755-a0e5-0426b18d8703.png)

**Comment by mattkindy** [comment]
I dug a bit into the generated code and noticed this section:

```ts
export const Thunder =
  (fn: FetchFunction) =>
  <O extends keyof typeof Ops, SCLR extends ScalarDefinition, R extends keyof ValueTypes = GenericOperation<O>>(
    operation: O,
    graphqlOptions?: ThunderGraphQLOptions<SCLR>,
  ) =>
  <Z extends ValueTypes[R]>(o: Z | ValueTypes[R], ops?: OperationOptions & { variables?: Record<string, unknown> }) =>
    fn(
      Zeus(operation, o, {
        operationOptions: ops,
        scalars: graphqlOptions?.scalars,
      }),
      ops?.variables,
    ).then((data) => {
      if (graphqlOptions?.scalars) {
        return decodeScalarsInResponse({
          response: data,
          initialOp: operation,
          initialZeusQuery: o as VType,
          returns: ReturnTypes,
          scalars: graphqlOptions.scalars,
          ops: Ops,
        });
      }
      return data;
    }) as Promise<InputType<GraphQLTypes[R], Z, SCLR>>;
```

It seemed to widen the acceptance input type a bit much... I could be wrong, but when I made the change to remove `Z` and simply continue using `ValueTypes[R]`, it seemed to work appropriately.

```ts
export const Thunder =
  (fn: FetchFunction) =>
  <O extends keyof typeof Ops, SCLR extends ScalarDefinition, R extends keyof ValueTypes = GenericOperation<O>>(
    operation: O,
    graphqlOptions?: ThunderGraphQLOptions<SCLR>,
  ) =>
  (o: ValueTypes[R], ops?: OperationOptions & { variables?: Record<string, unknown> }) =>
    fn(
      Zeus(operation, o, {
        operationOptions: ops,
        scalars: graphqlOptions?.scalars,
      }),
      ops?.variables,
    ).then((data) => {
      if (graphqlOptions?.scalars) {
        return decodeScalarsInResponse({
          response: data,
          initialOp: operation,
          initialZeusQuery: o as VType,
          returns: ReturnTypes,
          scalars: graphqlOptions.scalars,
          ops: Ops,
        });
      }
      return data;
    }) as Promise<InputType<GraphQLTypes[R], ValueTypes[R], SCLR>>;
```

**Comment by IceBlizz6** [comment]
Any updates on this?
If mattkindy's proposed changes works well then maybe look into making a pull request.
If not then would be nice if this could be prioritized.
As people have noted above, this is currently a big issue and the only blocker for my team to use it.

**Comment by mattkindy** [comment]
I will definitely be looking at this again. Have been slammed with other things -- am interested in helping resolving this.

**Comment by aexol** [comment]
This is a TS problem I will take a look soon. With latest versions of TS it should be possible to resolve it

**Comment by IceBlizz6** [comment]
@aexol  @mattkindy Did any of you get further on this?
My team needs to migrate soon, would be really great if this could get some love :)

**Comment by ValentinH** [comment]
In case it helps someone or the maintainers of this project, we fixed all the typings issues we had with zeus in our fork. Here are 2 PRs that could be useful to someone else:
- https://github.com/elba-security/graphql-zeus/pull/7
- https://github.com/elba-security/graphql-zeus/pull/8

Disclaimer: this is using graphql-zeus v4 as we never upgraded to v5. But I think that some type fixes could be applied to v5 as well.

**Comment by IceBlizz6** [comment]
That looks interesting, in case i want to try that fork then is it possible to get that from a package manager or do i need to compile it myself?

**Comment by ValentinH** [comment]
Yes it's available on npm:   @elba-security/codegen

See the other commits for small changes as there is no change log. Most changes are bug fixes, others are small adaptations to better fit with Hasura types (uuid or json for example).

**Comment by mattkindy** [comment]
> @aexol  @mattkindy Did any of you get further on this?
> My team needs to migrate soon, would be really great if this could get some love :)
> 

Afraid not. After @aexol dropped their comment, it coincided with this taking a backseat for us.

**Comment by IceBlizz6** [comment]
@aexol Have you looked further into this issue?

**Comment by IceBlizz6** [comment]
Is this fix on 5.3.2 ?
Cause i am testing it and it still seems to accept anything.
